File: sims/bsds_sims/bsd1.py
"""BSD1 run-file format: writer and reader.

Layout (spec docs/superpowers/specs/2026-08-18-basilisk-missions-stage1-design.md §4):

    bytes 0..7    magic ASCII "BSDS0001"
    bytes 8..11   uint32 LE header_length
    next          UTF-8 JSON header
    next          binary payload

The payload holds the time array (float64 LE) followed by each channel's
columns stored contiguously, column-major per channel (all samples of
component 0, then component 1, ...). Channel dtype is "f32" (float32 LE) or
"i16" (int16 LE, quantized). Quantization per component:

    offset = min(col)
    scale  = (max(col) - min(col)) / 65534   (scale = 1.0 when the range is 0)
    stored = round((x - offset) / scale) - 32767        # int16
    real   = (stored + 32767) * scale + offset

so round-trip error is bounded by scale/2. The TypeScript decoder in
web/src/bsd1.ts mirrors these formulas exactly — change both or neither.
"""

from __future__ import annotations

import json
import struct
from dataclasses import dataclass, field
from pathlib import Path

import numpy as np

MAGIC = b"BSDS0001"
_QSPAN = 65534  # int16 span used by quantization (symmetric about -32767..32767)
_QBIAS = 32767


@dataclass
class Channel:
    """One recorded signal: (n, components) float64 array plus metadata."""

    name: str
    data: np.ndarray
    unit: str = ""
    frame: str = ""
    dtype: str = "f32"  # "f32" | "i16"

    def __post_init__(self) -> None:
        arr = np.asarray(self.data, dtype=np.float64)
        if arr.ndim == 1:
            arr = arr[:, None]
        if arr.ndim != 2:
            raise ValueError(f"channel {self.name!r}: data must be 1-D or 2-D")
        if self.dtype not in ("f32", "i16"):
            raise ValueError(f"channel {self.name!r}: dtype must be f32 or i16")
        self.data = arr


@dataclass
class RunData:
    header: dict
    time_s: np.ndarray
    _channels: dict[str, np.ndarray] = field(default_factory=dict)

    def channel(self, name: str) -> np.ndarray | None:
        return self._channels.get(name)


def _quantize(col: np.ndarray) -> tuple[np.ndarray, float, float]:
    lo = float(col.min())
    hi = float(col.max())
    scale = (hi - lo) / _QSPAN if hi > lo else 1.0
    stored = np.round((col - lo) / scale).astype(np.int64) - _QBIAS
    return stored.astype("<i2"), scale, lo


def _dequantize(stored: np.ndarray, scale: float, offset: float) -> np.ndarray:
    return (stored.astype(np.float64) + _QBIAS) * scale + offset


def write_run(
    path: str | Path,
    *,
    scenario: str,
    run: str,
    title: str,
    epoch: str,
    params: dict,
    time_s: np.ndarray,
    channels: list[Channel],
    bodies: list[dict],
    metrics: dict,
) -> int:
    """Write one run file; returns total bytes written."""
    t = np.asarray(time_s, dtype=np.float64)
    if t.ndim != 1 or len(t) < 2:
        raise ValueError("time_s must be a 1-D array with at least 2 samples")
    if np.any(np.diff(t) <= 0):
        raise ValueError("time_s must be strictly increasing")
    n = len(t)

    payload = bytearray()
    payload += t.astype("<f8").tobytes()
    time_info = {"dtype": "f64", "unit": "s", "byte_offset": 0, "byte_length": n * 8}

    channel_infos = []
    for ch in channels:
        if ch.data.shape[0] != n:
            raise ValueError(f"channel {ch.name!r}: {ch.data.shape[0]} samples, expected {n}")
        comps = ch.data.shape[1]
        info: dict = {
            "name": ch.name,
            "components": comps,
            "unit": ch.unit,
            "frame": ch.frame,
            "dtype": ch.dtype,
            "byte_offset": len(payload),
        }
        if ch.dtype == "f32":
            for c in range(comps):
                payload += ch.data[:, c].astype("<f4").tobytes()
            info["byte_length"] = n * 4 * comps
        else:
            scales, offsets = [], []
            for c in range(comps):
                stored, scale, offset = _quantize(ch.data[:, c])
                payload += stored.tobytes()
                scales.append(scale)
                offsets.append(offset)
            info["byte_length"] = n * 2 * comps
            info["scale"] = scales
            info["offset"] = offsets
        channel_infos.append(info)

    header = {
        "schema": 1,
        "scenario": scenario,
        "run": run,
        "title": title,
        "epoch": epoch,
        "params": params,
        "n": n,
        "bodies": bodies,
        "time": time_info,
        "channels": channel_infos,
        "metrics": metrics,
    }
    header_bytes = json.dumps(header, separators=(",", ":")).encode("utf-8")

    out = Path(path)
    with out.open("wb") as f:
        f.write(MAGIC)
        f.write(struct.pack("<I", len(header_bytes)))
        f.write(header_bytes)
        f.write(bytes(payload))
    return 8 + 4 + len(header_bytes) + len(payload)


def read_run(path: str | Path) -> RunData:
    raw = Path(path).read_bytes()
    if len(raw) < 12 or raw[:8] != MAGIC:
        raise ValueError("not a BSD1 file (bad magic)")
    (hlen,) = struct.unpack("<I", raw[8:12])
    if len(raw) < 12 + hlen:
        raise ValueError("truncated BSD1 file (header)")
    header = json.loads(raw[12 : 12 + hlen].decode("utf-8"))
    payload = memoryview(raw)[12 + hlen :]

    def _view(offset: int, length: int) -> memoryview:
        if offset + length > len(payload):
            raise ValueError("truncated BSD1 file (payload)")
        return payload[offset : offset + length]

    tinfo = header["time"]
    n = header["n"]
    time_s = np.frombuffer(_view(tinfo["byte_offset"], tinfo["byte_length"]), dtype="<f8").copy()
    if len(time_s) != n:
        raise ValueError("time length mismatch")

    channels: dict[str, np.ndarray] = {}
    for ch in header["channels"]:
        comps = ch["components"]
        buf = _view(ch["byte_offset"], ch["byte_length"])
        out = np.empty((n, comps), dtype=np.float64)
        if ch["dtype"] == "f32":
            flat = np.frombuffer(buf, dtype="<f4")
            for c in range(comps):
                out[:, c] = flat[c * n : (c + 1) * n]
        elif ch["dtype"] == "i16":
            flat = np.frombuffer(buf, dtype="<i2")
            for c in range(comps):
                out[:, c] = _dequantize(flat[c * n : (c + 1) * n], ch["scale"][c], ch["offset"][c])
        else:
            continue  # unknown dtype: ignore channel (forward compatibility)
        channels[ch["name"]] = out

    return RunData(header=header, time_s=time_s, _channels=channels)

File: web/src/bsd1.ts
/** BSD1 run-file decoder — mirrors sims/bsds_sims/bsd1.py exactly.
 *
 * Layout: 8-byte magic "BSDS0001" | uint32 LE header length | UTF-8 JSON
 * header | binary payload. Channel columns are contiguous column-major per
 * channel. i16 dequantization: real = (stored + 32767) * scale + offset.
 */

export interface ChannelInfo {
  name: string;
  components: number;
  unit: string;
  frame: string;
  dtype: "f32" | "i16";
  byte_offset: number;
  byte_length: number;
  scale?: number[];
  offset?: number[];
}

export interface RunHeader {
  schema: number;
  scenario: string;
  run: string;
  title: string;
  epoch: string;
  params: Record<string, unknown>;
  n: number;
  bodies: { name: string; mu: number; radius_km: number }[];
  time: { dtype: string; unit: string; byte_offset: number; byte_length: number };
  channels: ChannelInfo[];
  metrics: Record<string, unknown>;
  [key: string]: unknown;
}

export interface DecodedChannel {
  info: ChannelInfo;
  components: number;
  /** Column-major storage: all samples of component 0, then component 1, ... */
  data: Float64Array;
  component(i: number): Float64Array;
  /** Value of component c at sample k. */
  at(k: number, c: number): number;
}

export interface RunData {
  header: RunHeader;
  time: Float64Array;
  channel(name: string): DecodedChannel | null;
  channelNames(): string[];
}

const MAGIC = "BSDS0001";
const QBIAS = 32767;

export function decodeRun(buf: ArrayBuffer): RunData {
  const bytes = new Uint8Array(buf);
  if (bytes.length < 12) throw new Error("not a BSD1 file (too short)");
  const magic = new TextDecoder().decode(bytes.subarray(0, 8));
  if (magic !== MAGIC) throw new Error("not a BSD1 file (bad magic)");
  const hlen = new DataView(buf).getUint32(8, true);
  if (bytes.length < 12 + hlen) throw new Error("truncated BSD1 file (header)");
  const header = JSON.parse(new TextDecoder().decode(bytes.subarray(12, 12 + hlen))) as RunHeader;

  // Slice payload into a fresh, 0-based ArrayBuffer so internal offsets are
  // naturally aligned for typed-array views.
  const payload = buf.slice(12 + hlen);
  const view = (off: number, len: number): ArrayBuffer => {
    if (off + len > payload.byteLength) throw new Error("truncated BSD1 file (payload)");
    return payload.slice(off, off + len);
  };

  const n = header.n;
  const time = new Float64Array(view(header.time.byte_offset, header.time.byte_length));
  if (time.length !== n) throw new Error("time length mismatch");

  const channels = new Map<string, DecodedChannel>();
  for (const info of header.channels) {
    const comps = info.components;
    const data = new Float64Array(n * comps);
    if (info.dtype === "f32") {
      const flat = new Float32Array(view(info.byte_offset, info.byte_length));
      for (let i = 0; i < n * comps; i++) data[i] = flat[i];
    } else if (info.dtype === "i16") {
      const flat = new Int16Array(view(info.byte_offset, info.byte_length));
      for (let c = 0; c < comps; c++) {
        const scale = info.scale![c];
        const offset = info.offset![c];
        for (let k = 0; k < n; k++) {
          data[c * n + k] = (flat[c * n + k] + QBIAS) * scale + offset;
        }
      }
    } else {
      continue; // unknown dtype: ignore channel (forward compatibility)
    }
    channels.set(info.name, {
      info,
      components: comps,
      data,
      component(i: number) {
        return data.subarray(i * n, (i + 1) * n);
      },
      at(k: number, c: number) {
        return data[c * n + k];
      },
    });
  }

  return {
    header,
    time,
    channel: (name) => channels.get(name) ?? null,
    channelNames: () => [...channels.keys()],
  };
}

/** Binary search: index of the last sample with time[i] <= t, clamped to [0, n-1]. */
export function nearestIndex(time: Float64Array, t: number): number {
  if (t <= time[0]) return 0;
  const n = time.length;
  if (t >= time[n - 1]) return n - 1;
  let lo = 0;
  let hi = n - 1;
  while (hi - lo > 1) {
    const mid = (lo + hi) >> 1;
    if (time[mid] <= t) lo = mid;
    else hi = mid;
  }
  return t - time[lo] <= time[hi] - t ? lo : hi;
}
